### PR TITLE
DBZ-349 Better support for large append-only tables by making the snapshotting process restartable for MySQL, updated configuration

### PR DESCRIPTION
**Pull request by ainagy**
Updated the way this feature can be configured to use individual config properties for each table instead of a JSON config.

### CODE REVIEW

**Comment by ainagy** [comment]
This change allows one to override the default 'select * from ...' used during snapshotting a table with a custom select statement, which makes it possible to continue an interrupted snapshot process for append-only tables.
This is useful for large tables that take a very long time to snapshot and thus are prone to be interrupted for some reason during snapshotting.
Today I have finally refactored the patch to use the proposed configuration syntax, using a config property that lists those tables where the snapshotting query should be overridden, and separate config properties for each table.
Also added one more integration test (specifying overrides for multiple tables).

Btw. we are now using the feature successfully in our environment with several-TB log tables.

**Comment by jpechane** [comment]
LGTM, thanks a lot!

**Comment by ainagy** [comment]
@jpechane thanks for having a look! Is there anything I can help with or answer about the patch?

**Comment by jpechane** [comment]
@ainagy we are waiting for @gunnarmorling approval stamp - he should be able to look at this either by ent of this week ar at the beginning of the next one.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java [hunk 1]
```diff
@@ -789,6 +789,16 @@ public static EventDeserializationFailureHandlingMode parse(String value) {
                              + "'warn' the problematic event and its binlog position will be logged and the event will be skipped;"
                              + "'ignore' the problematic event will be skipped.");
 
+    public static final Field SNAPSHOT_SELECT_STATEMENT_OVERRIDES_BY_TABLE = Field.create("snapshot.select.statement.overrides")
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: What's the rationale for having this additional property besides the props with the actual select statements? Wouldn't those be sufficient to control this functionality?

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java [hunk 2]
```diff
@@ -195,6 +198,20 @@ public boolean useMinimalSnapshotLocking() {
         return config.getBoolean(MySqlConnectorConfig.SNAPSHOT_MINIMAL_LOCKING);
     }
 
+    public Optional<String> getSnapshotSelectOverride(String tableId) {
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlTaskContext.java]: How about encapsulating the entire select statement determination in this method? It'd either return the default or the override if present.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlConnectorConfig.java]: Ok, realizing that @jpechane had been requesting this. Works for me :)

**Comment by gunnarmorling** [comment]
> Btw. we are now using the feature successfully in our environment with several-TB log tables.

Sweet, that's very valuable feedback. Thanks!

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks a lot, @ainagy! I've only added one more commit for improving the handling of the option a bit, it allows to get rid of the overrides map now after it has been used.

@jpechane, could you send a PR for the docs update for this one? Shouldn't be much. Thanks!

**Comment by jpechane** [comment]
Done - https://github.com/debezium/debezium.github.io/pull/95

**Pull request closed without merging**